### PR TITLE
feat(wms) Provide title to grouped layers (comma)

### PR DESCRIPTION
**Pull request by pelord**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines:https://github.com/infra-geo-ouverte/igo2/blob/master/.github/CONTRIBUTING.md#git-commit-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

image
The title of grouped wms layers are the name of the layer.
**What is the new behavior?**
For WMS grouped layers, with "optionsFromCapabilities" set true, the app will try to catch the title from the wms getcapabilities and provide it as the title.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications:

**Other information**:

### CODE REVIEW

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts [hunk 1]
```diff
@@ -27,5 +28,25 @@ export class LayerLegendComponent {
...
+  constructor(private capabilitiesService: CapabilitiesService) {}
+
+   computeItemTitle(item): string {
+    const layerOptions = this.layer.dataSource.options as any;
+    if (layerOptions.type === 'wms' && layerOptions.optionsFromCapabilities) {
+      let localLayerOptions = JSON.parse(JSON.stringify(layerOptions)); // to avoid to alter the original options.
+      layerOptions.params.layers.split(',').forEach(layer => {
+        if (layer === item.title) {
+          localLayerOptions.params.layers = layer;
+          this.capabilitiesService.getWMSOptions(localLayerOptions).subscribe(r => localLayerOptions = r);
```
Comment (cbourget) [inline packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: This runs asynchronously but your method doesn't return an observable.

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts [hunk 2]
```diff
@@ -27,5 +28,25 @@ export class LayerLegendComponent {
   }
   private _legend;
 
-  constructor() {}
+  constructor(private capabilitiesService: CapabilitiesService) {}
+
+   computeItemTitle(item): string {
+    const layerOptions = this.layer.dataSource.options as any;
+    if (layerOptions.type === 'wms' && layerOptions.optionsFromCapabilities) {
```
Comment (cbourget) [inline packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: That if statement contains the whole function body. Do an early return before entering that "if" if the conditions are not met.

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts [hunk 3]
```diff
@@ -27,5 +29,25 @@ export class LayerLegendComponent {
   }
   private _legend;
 
-  constructor() {}
+  constructor(private capabilitiesService: CapabilitiesService) {}
+
+   computeItemTitle(item): Observable<string> {
```
Comment (cbourget) [inline packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: Type item. I think it's a LayerLegend right?

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts (hunk 1 above)
Comment (cbourget) [reply packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: This still doesn't return an observable. You might not want to return an observable but, in that case, you can't subscribe to getWMSOptions (which makes a GetCapabilities) and assume that it'll complete in time for you to return localLayerOptions._layerOptionsFromCapabilities.title directly.

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts [hunk 4]
```diff
@@ -60,4 +61,21 @@ export class LayerLegendComponent implements OnInit, OnDestroy {
...
 
+  computeItemTitle(layerLegend): Observable<string> {
+    const layerOptions = this.layer.dataSource.options as any;
+    if (layerOptions.type !== 'wms') {
+      return of(layerLegend.title);
+    /*} else if (!layerOptions.optionsFromCapabilities) {
+      return of(layerLegend.title);*/
+    } else {
+      const layers = layerOptions.params.layers.split(',');
+      const localLayerOptions = JSON.parse(JSON.stringify(layerOptions)); // to avoid to alter the original options.
```
Comment (cbourget) [inline packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: I don't think this is still needed.

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts [hunk 5]
```diff
@@ -60,4 +61,21 @@ export class LayerLegendComponent implements OnInit, OnDestroy {
     this.legendItems$.next(legendItems);
   }
 
+  computeItemTitle(layerLegend): Observable<string> {
+    const layerOptions = this.layer.dataSource.options as any;
+    if (layerOptions.type !== 'wms') {
+      return of(layerLegend.title);
+    /*} else if (!layerOptions.optionsFromCapabilities) {
+      return of(layerLegend.title);*/
+    } else {
```
Comment (cbourget) [inline packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: No need for a else with an early return.

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts [hunk 6]
```diff
@@ -60,4 +61,21 @@ export class LayerLegendComponent implements OnInit, OnDestroy {
     this.legendItems$.next(legendItems);
   }
 
+  computeItemTitle(layerLegend): Observable<string> {
+    const layerOptions = this.layer.dataSource.options as any;
+    if (layerOptions.type !== 'wms') {
+      return of(layerLegend.title);
+    /*} else if (!layerOptions.optionsFromCapabilities) {
```
Comment (cbourget) [inline packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: Remove comment.

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts [hunk 7]
```diff
@@ -60,4 +61,21 @@ export class LayerLegendComponent implements OnInit, OnDestroy {
...
+      return of(layerLegend.title);*/
+    } else {
+      const layers = layerOptions.params.layers.split(',');
+      const localLayerOptions = JSON.parse(JSON.stringify(layerOptions)); // to avoid to alter the original options.
+      localLayerOptions.params.layers = layers.find(layer => layer === layerLegend.title);
+      return this.capabilitiesService
+        .getWMSOptions(localLayerOptions)
+        .pipe(map(wmsDataSourceOptions => {
+          return (wmsDataSourceOptions as any)._layerOptionsFromCapabilities.title;
```
Comment (cbourget) [inline packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: Is casting as any absolutely needed?

Comment (pelord) [reply packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: as in layer.service.ts, line 62.

Comment (pelord) [reply packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: if not :

Property '_layerOptionsFromCapabilities' does not exist on type 'WMSDataSourceOptions'.ts

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts (hunk 4 above)
Comment (pelord) [reply packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: Yes it is. It's the only way to parse the correct layer without altering the original one.

File: packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts (hunk 7 above)
Comment (cbourget) [reply packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: Should _layerOptionsFromCapabilities exist on WMSDataSourceOptions then?

Comment (pelord) [reply packages/geo/src/lib/layer/layer-legend/layer-legend.component.ts]: Added to interface.

**Pull request merged**